packages/dds/tree/src/shared-tree/treeApi.ts

```diff
@@ -0,0 +1,109 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { assert } from "@fluidframework/core-utils";
+import { fail } from "../util/index.js";
+import { Context } from "../feature-libraries/index.js";
+import {
+	TreeNode,
+	SimpleTreeApi,
+	TreeView,
+	simpleTreeApi,
+	getFlexNode,
+} from "../simple-tree/index.js";
+import { SchematizingSimpleTreeView } from "./schematizingTreeView.js";
+import { TreeCheckout } from "./treeCheckout.js";
+import { contextToTreeView } from "./treeView.js";
+
+/**
+ * Provides various functions for interacting with {@link TreeNode}s.
+ * @public
+ */
+export interface TreeApi extends SimpleTreeApi {
+	/**
+	 * Apply one or more edits to the tree as a single atomic unit.
+	 * @param node - The node that will be passed to `transaction`.
+	 * This is typically the root node of the subtree that will be modified by the transaction.
+	 * @param transaction - The function to run as the body of the transaction.
+	 * This function is passed the provided `node`.
+	 * At any point during the transaction, the function may return the value `"rollback"` to abort the transaction and discard any changes it made so far.
+	 * @remarks
+	 * All of the changes in the transaction are applied synchronously and therefore no other changes (either from this client or from a remote client) can be interleaved with those changes.
+	 * Note that this is guaranteed by Fluid for any sequence of changes that are submitted synchronously, whether in a transaction or not.
+	 * However, using a transaction has the following additional consequences:
+	 * - If reverted (e.g. via an "undo" operation), all the changes in the transaction are reverted together.
+	 * - If any change in the transaction fails and must be discarded, then the entire transaction is discarded.
```

Comment:
Until we have constraints, I don't think "failure" is a concept that is meaningful to users.

Reply:
Aren't there reasons outside of constraints that a transaction might fail? E.g. a transaction that tries to move a node that is deleted. I was under the impression that that is a sort of "implicit constraint" and it will ~fail~ rollback the whole transaction, even if there's not an explicit node existence constraint. Is that not the case?

Reply:
In which case it's not really "rolling back anything", it's simply that the final squash commit tried to do something impossible, and I thought that we apply commits either all the way or none of the way, not halfway.

Reply:
Discussed offline. My impression was false - there are no "failure cases" for unconstrained commits. All possible changes succeed.